ls1intum/Artemis
# Sync feature toggle using Hazelcast map.

### Checklist

### Motivation and Context
Artemis has feature toggles that allow administrators to disable certain features of Artemis without having to restart the application server (currently, only programming exercises can be disabled).
The feature toggle state (whether a certain feature is enabled or disabled) was currently only stored in an Enum on the instance where the administrator was connected to when performing the change but was not distributed between all instances.
This leads to inconsistent states where a feature can be enabled on one instance but disabled on another.

This PR synchronizes the feature state between all instances.

### Description
Instead of the Enum that holds the state (which cannot easily be shared), a map is shared between all instances:
- Each feature is either enabled (true) or disabled (false)

### Steps for Testing
This PR can only really be tested if you are running at least two instances of Artemis (thus you should use TS2 for testing).
You can however also test on TS1/3 and just make sure nothing broke.

1. Log in to Artemis
2. Navigate to Server Administration >  Feature Toggles
3. Toggle the state of a feature: The change should immediately be effective to all users, even if they are connected to a different instance.

This can be tested on TS2 with multiple testers connected to different instances, just DM me (or anyone else with access to TS2) on Slack, so that we can verify that the testers are indeed connected to different instances.

### Test Coverage
**Server:**
- FeatureToggleService: 100%
